### PR TITLE
Added `Op`s for `np.linspace`, `np.logspace`, and `np.geomspace`

### LINKED ISSUE #768: Implement Op for `linspace` and similar numpy functions
The following numpy functions don't have to counterpart in Aesara:
* [linspace](https://numpy.org/doc/stable/reference/generated/numpy.linspace.html)
* [logspace](https://numpy.org/doc/stable/reference/generated/numpy.logspace.html)
* [geomspace](https://numpy.org/doc/stable/reference/generated/numpy.geomspace.html)

I am not familiar with all of them, so some might not make sense, or they might be low priority.
`linspace` would definitely be nice to have.

### PR DESCRIPTION
**Pull request by kc611**
Resolves #768

Adds `Op` counterparts for the following numpy functions:
- `np.linspace`
- `np.logspace`
- `np.geomspace`

### CODE REVIEW

**Comment by ricardoV94** [comment]
It should be possible to obtain gradients with respect to the endpoints but I have no idea ehat these would be. Jax seems to be able to do it...

**Comment by kc611** [comment]
Seems like the devs at `Jax` replaced direct calls to `np.linspace` by a custom `linspace` function which was differentiable w.r.t `start` and `stop` (not by building a gradient version of `linspace` itself but by splitting `linspace` into operations which were differentiable). I don't know how that could be done in Aesara.

Have a look at: https://github.com/google/jax/pull/1658

**Comment by ricardoV94** [comment]
Can we also compose it in terms of already implemented builtins?

If not we should be able to figure out the symbol grads. For instance for linspace with respect to start, I think it would be something like `gx * (1 - arange(steps-1)/steps-1)`

**Comment by ricardoV94** [comment]
Alternatively, linspace can easily be built from `arange` no?

**Comment by kc611** [comment]
> Alternatively, linspace (at least) can easily be built from arange no?

I just realized that. We should probably just make `linspace` an alias to `arange` rather than having a new `Op` entirely for that. Though I don't know about the other two.

**Comment by brandonwillard** [comment]
> Seems like the devs at `Jax` replaced direct calls to `np.linspace` by a custom `linspace` function which was differentiable w.r.t `start` and `stop` (not by building a gradient version of `linspace` itself but by splitting `linspace` into operations which were differentiable). I don't know how that could be done in Aesara.

That's actually how Theano did quite a few of its NumPy function implementations, and it's still the prefered way of doing things in Aesara.  For instance, see the opening to [this section in the documentation](https://aesara.readthedocs.io/en/latest/extending/creating_an_op.html).

In other words, **implement operations using existing `Op`s instead of creating new `Op`s**, when possible.

Not only does this approach prevent one from needing to implement an `Op.[grad, L_op, R_op]`, but it also opens up those operations to optimization, whereas the internal operations of a custom `Op` obviously can't be optimized (by us).

**Comment by kc611** [comment]
> Alternatively, linspace (at least) can easily be built from arange no?

It seems that `ARange` is limited in functionality:

https://github.com/aesara-devs/aesara/blob/80bfde1ee23f6239ae86cdaa42af9e383048b759/aesara/tensor/basic.py#L3249-L3251

Would it be better to have this as is, or should we expand the functionality of `ARange` ?

**Comment by ricardoV94** [comment]
> Ah no, I meant for cases like: 
> 
> ```
> start = np.array([5, 6])
> stop = np.array([[10, 10], [10, 10]])
> step = 25
> ```

I think my pseudo code would work fine for those

**Comment by kc611** [comment]
Ah yes, It does, sorry, guess I need a stronger coffee.

**Comment by ricardoV94** [comment]
> Ah yes, It does, sorry, guess I need a stronger coffee.

I edited my message like 15x times, so you're not alone xD

**Comment by ricardoV94** [comment]
logspace might be something like:

```python
def logspace(start, end, steps, base=10.0):
  return log(linspace(base**start, base**end, steps))
```

**Comment by ricardoV94** [comment]
NVM Numpy says it is:

```pythom
def logspace(start, end, steps, base):
  power(base, linspace(start, end steps))
```
https://numpy.org/doc/stable/reference/generated/numpy.logspace.html

Maybe my previous one was geomspace?

File: aesara/tensor/extra_ops.py [hunk 1]
```diff
@@ -1625,6 +1625,29 @@ def infer_shape(self, fgraph, node, ins_shapes):
...
+    start = at.as_tensor_variable(start)
+    end = at.as_tensor_variable(end)
+    return base ** linspace(start, end, steps)
+
+
+def linspace(start, end, steps):
+    start = at.as_tensor_variable(start)
+    end = at.as_tensor_variable(end)
+    arr = at.arange(steps)
+    arr = arr.dimshuffle([0] + ["x"] * max(start.ndim, end.ndim))
```
Comment (ricardoV94) [inline aesara/tensor/extra_ops.py:1646]: This one is a bit hard to read. Is it equivalent to  the following?


```suggestion
    arr = at.shape_padright(max(start.ndim, end.ndim))
```

**Review by ricardoV94 — COMMENTED** [review]
Looks great! Small suggestion/question below

File: aesara/tensor/extra_ops.py (hunk 1 above)
Comment (ricardoV94) [reply aesara/tensor/extra_ops.py]: Maybe it was padleft?

**Review by brandonwillard — DISMISSED** [review]

**Review by brandonwillard — APPROVED** [review]

**Comment by brandonwillard** [comment]
Agh, looks like a merge commit got in there somehow.

**Pull request merged**